packages/template-retail-react-app/app/partials/product-view/index.jsx

```diff
@@ -76,175 +76,366 @@ const ButtonWithRegistration = withRegistration(Button)
+                if (!hasValidSelection) {
+                    return null
+                }
 
-        return buttons
-    }
+                if (!addToCart && !updateCart) return null
+                if (updateCart) {
+                    await updateCart(variant, quantity)
+                    return
+                }
+                const itemsAdded = await addToCart(variant, quantity)
+
+                onAddToCartModalOpen({
+                    product,
+                    itemsAdded
+                })
+            }
+
+            const handleWishlistItem = async () => {
+                if (!updateWishlist && !addToWishlist) return null
+                if (updateWishlist) {
+                    updateWishlist(product, variant, quantity)
+                    return
+                }
+                addToWishlist(product, variant, quantity)
+            }
+
+            if (addToCart || updateCart) {
+                buttons.push(
+                    <Button
+                        key="cart-button"
+                        onClick={handleCartItem}
+                        disabled={showInventoryMessage}
+                        width="100%"
+                        variant="solid"
+                        marginBottom={4}
+                    >
+                        {updateCart
+                            ? buttonText.update
+                            : isProductASet
+                            ? buttonText.addSetToCart
+                            : buttonText.addToCart}
+                    </Button>
+                )
+            }
+
+            if (addToWishlist || updateWishlist) {
+                buttons.push(
+                    <ButtonWithRegistration
+                        key="wishlist-button"
+                        onClick={handleWishlistItem}
+                        disabled={isWishlistLoading || !canAddToWishlist}
+                        isLoading={isWishlistLoading}
+                        width="100%"
+                        variant="outline"
+                        marginBottom={4}
+                    >
+                        {updateWishlist
+                            ? buttonText.update
+                            : isProductASet
+                            ? buttonText.addSetToWishlist
+                            : buttonText.addToWishlist}
+                    </ButtonWithRegistration>
+                )
+            }
 
-    useEffect(() => {
-        if (isAddToCartModalOpen) {
-            onAddToCartModalClose()
+            return buttons
         }
-    }, [location.pathname])
 
-    useEffect(() => {
-        if (canOrder) {
-            toggleShowOptionsMessage(false)
+        // Bind the reference with our `scope` that includes the internal validate function for this component.
+        // Other values can be added to this scope as required.
+        if (typeof ref === 'function') {
+            ref = ref.bind({validateAttributeSelection: validateAndShowError})
         }
```

Comment:
This will be called on every render.. making the function binding happen multiple times. Not sure what the end result will be (like a Russian doll?), but can we do the binding only once?

Reply:
Surprisingly it doesn't work if was to put this in a `useEffect` hook. Still looking into alternatives. Do you consider this a blocker?

Reply:
I was able to use an ref to only do this binding once 👍

Reply:
Binding a bound function has no effect; the first `this` context provided by `.bind` (or from being an arrow function) is used.